src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -765,15 +777,37 @@ ExecuteSingleTask(QueryDesc *queryDesc, Task *task,
 		bool queryOK = false;
 		bool failOnError = false;
 		int64 currentAffectedTupleCount = 0;
-		PGconn *connection = GetConnectionForPlacement(taskPlacement,
-													   isModificationQuery);
+		int connectionFlags = SESSION_LIFESPAN;
+		MultiConnection *multiConnection = GetNodeConnection(connectionFlags,
+															 taskPlacement->nodeName,
+															 taskPlacement->nodePort);
+		PGconn *connection = multiConnection->pgConn;
+
+		if (taskRequiresTwoPhaseCommit)
+		{
+			RemoteTransactionBeginIfNecessary(multiConnection);
+			MarkRemoteTransactionCritical(multiConnection);
+		}
```

Comment:
Oh, and I suspect it'd be better to move the RemoteTransactionBeginIfNecessary one layer up, and establish connections for all placements at once. That way we can avoid paying the latency price for each node separately. That becomes fairly important if you have like 10 nodes.